src/net_processing.cpp

```diff
@@ -2961,14 +2961,16 @@ static bool SendRejectsAndCheckIfBanned(CNode* pnode, CConnman* connman, bool en
             LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode->addr.ToString());
         else if (pnode->m_manual_connection)
             LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode->addr.ToString());
-        else {
-            pnode->fDisconnect = true;
-            if (pnode->addr.IsLocal())
-                LogPrintf("Warning: not banning local peer %s!\n", pnode->addr.ToString());
-            else
-            {
-                connman->Ban(pnode->addr, BanReasonNodeMisbehaving);
+        else if (pnode->addr.IsLocal()) {
+            // Disconnect but don't ban _this_ local node
+            LogPrintf("Warning: disconnecting but not banning local peer %s!\n", pnode->addr.ToString());
+            connman->DisconnectNode(pnode->GetId());
```

Comment:
nit: could have a `CNode*` version of `DisconnectNode` to avoid the lookup

Reply:
I'm going to revert to doing
```
pnode->fDisconnect = true;
```
so as to avoid the lookup, then in a followup PR we can make `fDisconnect` private, and have a `Disconnect` method on `CNode`.